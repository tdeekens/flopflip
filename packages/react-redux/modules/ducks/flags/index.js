export {
  default as flagsReducer,
  createReducer as createFlagsReducer,
} from './flags';
export { UPDATE_FLAGS, updateFlags, selectFlag, selectFlags } from './flags';
