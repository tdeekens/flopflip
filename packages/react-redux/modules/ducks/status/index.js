export { default as statusReducer } from './status';
export { UPDATE_STATUS, updateStatus } from './status';
