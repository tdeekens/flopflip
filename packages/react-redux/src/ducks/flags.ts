import { isNil } from '@flopflip/react';
import type {
  TAdapterIdentifiers,
  TFlagName,
  TFlagVariation,
  TFlagsChange,
  TFlagsContext,
} from '@flopflip/types';
import { type PayloadAction, createSlice } from '@reduxjs/toolkit';

import { STATE_SLICE } from '../constants';
import type { TState } from '../types';

const initialState: TFlagsContext = { memory: {} };

const flagsSlice = createSlice({
  name: 'flags',
  initialState,
  reducers: {
    updateFlags: {
      reducer(
        state,
        action: PayloadAction<
          TFlagsChange & { adapterIdentifiers: TAdapterIdentifiers[] }
        >
      ) {
        if (action.payload.id) {
          state[action.payload.id] = {
            ...state[action.payload.id],
            ...action.payload.flags,
          };
          return;
        }

        for (const adapterId of action.payload.adapterIdentifiers) {
          state[adapterId] = {
            ...state[adapterId],
            ...action.payload.flags,
          };
        }
      },
      prepare(
        flagsChange: TFlagsChange,
        adapterIdentifiers: TAdapterIdentifiers[]
      ) {
        return {
          payload: { ...flagsChange, adapterIdentifiers },
        };
      },
    },
  },
});

export const { updateFlags } = flagsSlice.actions;
export const reducer = flagsSlice.reducer;

export const createReducer = (preloadedState: TFlagsContext = initialState) => {
  // biome-ignore lint/style/useDefaultParameterLast: <explanation>
  return (state = preloadedState, action: ReturnType<typeof updateFlags>) =>
    reducer(state, action);
};

export const selectFlags = () => (state: TState) =>
  state[STATE_SLICE].flags ?? {};

export const selectFlag =
  (flagName: TFlagName, adapterIdentifiers: TAdapterIdentifiers[]) =>
  (state: TState): TFlagVariation => {
    const allFlags = selectFlags()(state);
    let foundFlagVariation: TFlagVariation = false;

    for (const adapterId of adapterIdentifiers) {
      const flagValue = allFlags[adapterId]?.[flagName];
      if (!isNil(flagValue)) {
        foundFlagVariation = flagValue;
      }
    }

    return foundFlagVariation;
  };
