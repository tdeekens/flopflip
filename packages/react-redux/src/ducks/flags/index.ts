export {
  createReducer as createFlagsReducer,
  default as flagsReducer,
} from './flags';
export { selectFlag, selectFlags, UPDATE_FLAGS, updateFlags } from './flags';
