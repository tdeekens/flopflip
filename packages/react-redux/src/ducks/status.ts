import { selectAdapterConfigurationStatus } from '@flopflip/react';
import type {
  TAdapterIdentifiers,
  TAdapterStatusChange,
  TAdaptersStatus,
} from '@flopflip/types';
import {
  type PayloadAction,
  createSelector,
  createSlice,
} from '@reduxjs/toolkit';

import { STATE_SLICE } from '../constants';
import type { TState } from '../types';

const initialState: TAdaptersStatus = {};

const statusSlice = createSlice({
  name: 'status',
  initialState,
  reducers: {
    updateStatus: {
      reducer(
        state,
        action: PayloadAction<
          TAdapterStatusChange & { adapterIdentifiers: TAdapterIdentifiers[] }
        >
      ) {
        if (action.payload.id) {
          state[action.payload.id] = {
            ...state[action.payload.id],
            ...action.payload.status,
          };
          return;
        }

        for (const adapterInterfaceIdentifier of action.payload
          .adapterIdentifiers) {
          state[adapterInterfaceIdentifier] = {
            ...state[adapterInterfaceIdentifier],
            ...action.payload.status,
          };
        }
      },
      prepare(
        statusChange: TAdapterStatusChange,
        adapterIdentifiers: TAdapterIdentifiers[]
      ) {
        return {
          payload: { ...statusChange, adapterIdentifiers },
        };
      },
    },
  },
});

export const { updateStatus } = statusSlice.actions;
export const reducer = statusSlice.reducer;

// Selectors
type TSelectStatusArgs = {
  adapterIdentifiers?: TAdapterIdentifiers[];
};

export const selectStatus =
  ({ adapterIdentifiers }: TSelectStatusArgs = {}) =>
  (state: TState) => {
    const { status } = state[STATE_SLICE];
    return selectAdapterConfigurationStatus(status, adapterIdentifiers);
  };
