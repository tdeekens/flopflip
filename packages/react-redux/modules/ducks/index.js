// @flow

import type { FlagName, FlagVariation, Flag, Flags } from '@flopflip/types';

import { combineReducers } from 'redux';
import { flagsReducer, createFlagsReducer } from './flags';
import { statusReducer } from './status';

export { updateStatus, UPDATE_STATUS } from './status';
export { updateFlags, UPDATE_FLAGS, selectFlag, selectFlags } from './flags';

export const flopflipReducer = combineReducers({
  flags: flagsReducer,
  status: statusReducer,
});
export const createFlopflipReducer = (preloadedState: Flags = {}) =>
  combineReducers({
    flags: createFlagsReducer(preloadedState),
    status: statusReducer,
  });
