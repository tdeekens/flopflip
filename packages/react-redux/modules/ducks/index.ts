import { Reducer } from 'redux';
import { combineReducers } from 'redux';
import { Flags } from '@flopflip/types';
import { UpdateFlagsAction } from './flags/types';
import { UpdateStatusAction } from './status/types';
import { flagsReducer, createFlagsReducer } from './flags';
import { statusReducer } from './status';

type Actions = UpdateFlagsAction & UpdateStatusAction;

export { updateStatus, UPDATE_STATUS } from './status';
export { updateFlags, UPDATE_FLAGS, selectFlag, selectFlags } from './flags';

export const flopflipReducer = combineReducers<any, Actions>({
  flags: flagsReducer,
  status: statusReducer,
});
export const createFlopflipReducer = (
  preloadedState: Flags = {}
): Reducer<any, Actions> =>
  combineReducers<any, Actions>({
    flags: createFlagsReducer(preloadedState),
    status: statusReducer,
  });
