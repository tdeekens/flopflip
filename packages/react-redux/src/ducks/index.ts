import type { TFlagsContext } from '@flopflip/types';
import { combineReducers } from 'redux';

import { createFlagsReducer, flagsReducer } from './flags';
import type { TUpdateFlagsAction } from './flags/types';
import { statusReducer } from './status';
import type { TUpdateStatusAction } from './status/types';

type Actions = TUpdateFlagsAction & TUpdateStatusAction;

export { selectFlag, selectFlags, UPDATE_FLAGS, updateFlags } from './flags';
export { UPDATE_STATUS, updateStatus } from './status';

export const flopflipReducer = combineReducers<any, Actions>({
  flags: flagsReducer,
  status: statusReducer,
});
export const createFlopflipReducer = (
  preloadedState: TFlagsContext = { memory: {} }
) =>
  combineReducers<any, Actions>({
    flags: createFlagsReducer(preloadedState),
    status: statusReducer,
  });
