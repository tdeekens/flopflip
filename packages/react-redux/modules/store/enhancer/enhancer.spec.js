import { updateFlags, updateStatus } from '../../ducks';
import createFlopFlipEnhancer from './enhancer';

const adapterArgs = {
  clientSideId: '123-abc',
  user: { key: 'foo-user' },
};
const adapter = {
  configure: jest.fn(),
  reconfigure: jest.fn(),
};

describe('when creating enhancer', () => {
  let enhancer;
  beforeEach(() => {
    enhancer = createFlopFlipEnhancer(adapter, adapterArgs);
  });

  describe('with enhanced store', () => {
    let dispatch;

    beforeEach(() => {
      dispatch = jest.fn();

      const getState = () => ({});
      const next = jest.fn(() => ({ getState, dispatch }));
      const args = [''];

      enhancer(next)(args);
    });

    it('should invoke `configure` on `adapter`', () => {
      expect(adapter.configure).toHaveBeenCalled();
    });

    it('should invoke `configure` on `adapter` with `adapterArgs`', () => {
      expect(adapter.configure).toHaveBeenCalledWith(
        expect.objectContaining(adapterArgs)
      );
    });

    it('should invoke `configure` on `adapter` with `onFlagsStateChange`', () => {
      expect(adapter.configure).toHaveBeenCalledWith(
        expect.objectContaining({
          onFlagsStateChange: expect.any(Function),
        })
      );
    });

    it('should invoke `configure` on `adapter` with `onStatusStateChange`', () => {
      expect(adapter.configure).toHaveBeenCalledWith(
        expect.objectContaining({
          onStatusStateChange: expect.any(Function),
        })
      );
    });

    describe('when invoking  `onFlagsStateChange`', () => {
      let nextFlags = {
        foo: true,
      };

      beforeEach(() => {
        const { onFlagsStateChange } = adapter.configure.mock.calls[
          adapter.configure.mock.calls.length - 1
        ][0];

        onFlagsStateChange(nextFlags);
      });

      it('should match snapshot', () => {
        expect(dispatch).toMatchSnapshot();
      });

      it('should invoke `dispatch`', () => {
        expect(dispatch).toHaveBeenCalled();
      });

      it('should invoke `dispatch` with `updateFlags`', () => {
        expect(dispatch).toHaveBeenCalledWith(updateFlags(nextFlags));
      });
    });

    describe('when invoking  `onStatusStateChange`', () => {
      let nextStatus = {
        isReady: true,
      };

      beforeEach(() => {
        const { onStatusStateChange } = adapter.configure.mock.calls[
          adapter.configure.mock.calls.length - 1
        ][0];

        onStatusStateChange(nextStatus);
      });

      it('should match snapshot', () => {
        expect(dispatch).toMatchSnapshot();
      });

      it('should invoke `dispatch`', () => {
        expect(dispatch).toHaveBeenCalled();
      });

      it('should invoke `dispatch` with `updateStatus`', () => {
        expect(dispatch).toHaveBeenCalledWith(updateStatus(nextStatus));
      });
    });
  });
});
