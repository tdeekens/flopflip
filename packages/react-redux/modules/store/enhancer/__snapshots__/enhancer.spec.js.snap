// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`when creating enhancer with enhanced store when invoking  \`onFlagsStateChange\` should match snapshot 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "payload": Object {
          "foo": true,
        },
        "type": "@flopflip/flags/update",
      },
    ],
  ],
  "results": Array [
    Object {
      "isThrow": false,
      "value": undefined,
    },
  ],
}
`;

exports[`when creating enhancer with enhanced store when invoking  \`onStatusStateChange\` should match snapshot 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "payload": Object {
          "isReady": true,
        },
        "type": "@flopflip/status/update",
      },
    ],
  ],
  "results": Array [
    Object {
      "isThrow": false,
      "value": undefined,
    },
  ],
}
`;
